[php] use http_build_query for deepObject support

nadar opened the pull request:
This fixes #11222 

The [Query::build()](https://github.com/guzzle/psr7/blob/master/src/Query.php#L71) function does not support deepObjects in query params, but the [OpenAPI defintion](https://swagger.io/docs/specification/serialization/) does. 

> `/users?id[role]=admin&id[firstName]=Alex` an example from the docs.

I have replaced the `Query::build()` function with PHP's internal [http_build_query](https://www.php.net/http_build_query) function. I have adjust PHP default encoding to match the default encoding from `Query::build()` which is `PHP_QUERY_RFC3986`.

This change will only have an effect for the query string of the URL, encoding POST bodys is still done by `Query::build()`.

ybelenko commented:
cc @jebentier, @dkarlovi, @mandrean, @jfastnacht, @ackintosh, @renepardon

nadar commented:
Is there something i can improve to fix bitrise?

wing328 commented:
For the bitrise CI failure, please ignore it.

wing328 commented:
@nadar thanks for the PR. Can we add some tests in https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/php/OpenAPIClient-php/tests/ParametersTest.php ?

nadar commented:
Sure!

nadar commented:
@wing328 I have to admit i am not sure what kind of test you expect. As i have changed the default behavior in tests (i have modified all tests apis with the new code), if they are not failing there should be no BC break. I can add a very simple nested object tested for URLs, i will do that, but not sure if it helps, or makes anything more stable :-) but i'll do it.

ybelenko commented:
I'm taking care of it. Please don't merge or push anything. Gonna push commits soon.

nadar commented:
@ybelenko thanks for your help on this :+1: i have at least added a tests which shows that the default behavior is the same for query param generating.

ybelenko commented:
By the way, why don't we ask for deepObject support at Guzzle repo?

nadar commented:
Sure you can :+1: The be honest i have no idea what that function is made for, seems to be a sort of "ensure only scalar values are provided", but i assume there is a reason to have that function, otherwise they would just use `http_build_query` instead. So this might be a huge "BC break" for the guzzle eco system. This also might take very long to introduce deep object params here. So i would suggest, to ask them by issue, but in the meantime merge this PR, as we are really looking forward to have this option.

nadar commented:
@ybelenko @wing328  Any news regarding this PR? Is there something i can do? It would be really nice to have that feature soon. Thanks

wing328 commented:
@nadar thanks for adding the tests. It looks good from my side and happy to see this PR gets merged.

ybelenko commented:
Doesn't look good to me…

nadar commented:
@ybelenko i see, those data is generated diff checked, now i can more understand how the test works.

ybelenko commented:
I used examples from [OpenAPI-Specification Style Examples](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#style-examples) to write fixtures. Also used doc from Swagger [Parameter Serialization](https://swagger.io/docs/specification/serialization/). Also checked that SwaggerEditor gives the same queries with fake spec file. Spent almost 4 days on that features, Jesus.

```console
$ vendor/bin/phpunit tests/ParametersTest.php

PHPUnit 9.5.10 by Sebastian Bergmann and contributors.

....                                                                4 / 4 (100%)

Time: 00:00.054, Memory: 6.00 MB

OK (4 tests, 6 assertions)
```
```console
$ vendor/bin/phpunit tests/ObjectSerializerTest.php

PHPUnit 9.5.10 by Sebastian Bergmann and contributors.

...............................................                   47 / 47 (100%)

Time: 00:00.023, Memory: 6.00 MB

OK (47 tests, 57 assertions)

nadar commented:
Thanks for the work. I have admit i don't understand why things must contain such complex code which needs to be maintained and understand by users instead of `http_build_query`... but up to you guys

> // nested arrays doesn't work, but it might help in future

What do you mean by this? Thats a valid deepObject requirement if so. 

```
{ "filter": { "or": [ {"name":"John"}, {"email":"john@doe.com"} ] } } 
```

should be encoded to:

```
filter[or][0][name]=John&filter[or][1][email]=john@doe.com
```

or in php:

```php
<?php

$filter = [
 'filter' => [
   'or' => [
    ['name' => 'John'],
    ['email' => 'john@doe.com']
   ]
 ]
];

var_dump(urldecode(http_build_query($filter)));
```

could you add that to your test case?

ybelenko commented:
> or in php:
> 
> ```
> <?php
> 
> $filter = [
>  'filter' => [
>    'or' => [
>     ['name' => 'John'],
>     ['email' => 'john@doe.com']
>    ]
>  ]
> ];
> 
> var_dump(urldecode(http_build_query($filter)));
> ```
> 
> could you add that to your test case?

Damn, it doesn't work. I need more time then. However `http_build_query($value, '', '&', \PHP_QUERY_RFC3986)` breaks all serializations with `explode: true` option.

The difference of `Query::build` method:
> This function does not modify the provided keys when an array is encountered (like `http_build_query()` would).

I don't know exactly why we use `Query::build` instead of `http_build_query`, but I guess it was there on purpose.

ybelenko commented:
The main issue is `http_build_query` won't serialise  `$arr = ["color" => ["blue", "black", "brown"]]` into `"color=blue&color=black&color=brown"` when parameter style is `form` and explode is `true`. I also cannot execute different build calls for each query parameter because I don't know how to merge them after.

ybelenko commented:
Well, it becomes even more tricky. 
My input:

```yaml
  /pet/findByStatus:
    get:
      tags:
      - pet
      summary: Finds Pets by status
      description: Multiple status values can be provided with comma separated strings
      operationId: findPetsByStatus
      parameters:
      - name: filter
        in: query
        style: deepObject
        required: true
        schema:
          type: object
          properties:
            or:
              type: array
              items:
                type: object
                additionalProperties:
                  type: string
```
with parameter value
```json
{
    "or": [
        {
            "name": "John",
            "email": "john@doe.com"
        },
        {
            "name": "Bob",
            "email": "bob@doe.com"
        }
    ]
}
```
I checked nested object serialization in [Online Swagger Editor](https://editor.swagger.io/) it gives a weird string(encoded and non-encoded at the same time, also array index missing):
```url
?filter%5Bor%5D[name]=John&filter%5Bor%5D[email]=john@doe.com
&filter%5Bor%5D[name]=Bob&filter%5Bor%5D[email]=bob@doe.com
```
then I checked with [VsCode extension - OpenAPI (Swagger) Editor](https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi) output is the same. So it also skips indexes.

@wing328 what do you think, which url string we should consider correct for `style: deepObject`?(Nadar or Swagger output or maybe something else)

@nadar I can push commit with your test case and you can try something yourself. I don't think 5.4.0 will be published this January, so we have some time anyway.

nadar commented:
@ybelenko @wing328 thanks for the research, thanks for taking time. Open-Source is only as strong as its contributors, i totally appreciate that!

The javascript sdk works perfectly with at least this example:

```
({filter: [{'tags.alias':'food'}]}
```

this will transform the get request to `?filter[tags.alias]=food`. So the PHP sdk should be able to achieve the same, otherwise the idea of openapi specifications is a bit gone :smile: i think we all agree on this. Does your currently solution support inputs like:

```
['tags.alias' => 'food']
```
instead of
```
['tags' => ['alias' => 'food']]
`` 
then we are good to i would say.

wing328 commented:
> this will transform the get request to ?filter[tags.alias]=food. So the PHP sdk should be able to achieve the same,

Agree 👍

ybelenko commented:
@nadar I'm really sorry, I've missed 5.4.0 released. Gonna take care of this PR in a few weeks.

nadar commented:
@ybelenko would be really nice to have this feature soon, otherwise the sdk is actually usless as there are loot of deepObject parameters :smiling_face_with_tear: if there is something i can help, please let me know. I will take time to do so, as its a very important topic for us

ybelenko commented:
> @ybelenko would be really nice to have this feature soon, otherwise the sdk is actually usless as there are loot of deepObject parameters :smiling_face_with_tear: if there is something i can help, please let me know. I will take time to do so, as its a very important topic for us

Gonna handle that next Monday. Your welcome to provide more complex examples of query serialization for tests. It can save me some time. We didn't even find what output is correct and what is not. It's not so obvious. OpenApi manual doesn't cover all cases, Swagger manual describes it differently, live Swagger editor produces weird strings.
Let's approve examples first, then I can fix implementation.

Default `http_build_query` works great with deepObject, but it doesn't with other serialization styles, I need to find some workarounds for users who wanna use different styles in the same request.

The final code will be ugly, but if it passes tests it should do the job. Sadly there is no place for an elegant code in codegen project, OpenApi spec is so complicated so implementation requires hacks and patches here and there. I already noticed it with sdk models(components), you start with cute classes like from the book and then spec says "Model can be primitive type, consider your User class as a string or Boolean" and the code of your model turns into crap immediately.

nadar commented:
>  I already noticed it with sdk models(components), you start with cute classes like from the book and then spec says "Model can be primitive type, consider your User class as a string or Boolean" and the code of your model turns into crap immediately.

:laughing:

ybelenko commented:
@nadar Btw if you need this fix asap why don't you overwrite the template? Every time I don't like something in built-in template or when it doesn't fit my needs  I just overwrite it. If you need `deepObject` and not going to use other serialization styles(csv, tsv, pipe etc.) you can go with `http_build_query()` easy.

nadar commented:
@ybelenko yes i am aware of the option to override things thanks, but this not a really a solution for the problem (and also not for us, as we generate several sdks dynamically on server side for multiple customers).

ybelenko commented:
@nadar How about `queryBuilder` generator option (like `--additional-properties=queryBuilder=guzzle|http_build_query`, default is guzzle)? Will it satisfy you?

If you need it so badly and you don't want to overwrite template I can add that generator option. Usually I don't add options which modifies single line of code.

nadar commented:
Maybe it should be `--additional-properties=deepObjectSupport=yes` so its decoupled from the function and might change other things in future.

ybelenko commented:
@nadar Please, check it out. I've added anonymous function to flatten query array so it solves Guzzle error. Should work with nested arrays now since all tests are passed.

Also wrote small static function to wrap query build, we can migrate to `http_build_query` with one line of code in future:
https://github.com/OpenAPITools/openapi-generator/blob/920cfe5463e6580b293e3e8a35bb201fb0728438/samples/client/petstore/php/OpenAPIClient-php/lib/ObjectSerializer.php#L452-L471

```console
vendor/bin/phpunit tests/ObjectSerializerTest.php
PHPUnit 9.5.18

..................................................                50 / 50 (100%)

Time: 00:00.059, Memory: 6.00 MB

OK (50 tests, 60 assertions)
```

nadar commented:
So you could fix the original issue without the `--additional-properties` option? That's great news! So my provided test case works now out of the box?

ybelenko commented:
> So you could fix the original issue without the `--additional-properties` option? That's great news! So my provided test case works now out of the box? 

I think so, at least minimal examples list works out of the box, including yours. I'm scared to test it with crazy fixtures(nested arrays of unexpected values) 😅 😅 😅

ybelenko commented:
@nadar I'm pretty satisfied with the result here. I wait for your feedback, hope you can check it locally then I can merge.

nadar commented:
i will do a test tomorrow, but if my provided example works then i am totally happy! 

```
['tags' => ['alias' => 'food']]
```

should be evaluated as

```
?filter[tags.alias]=food
```

nadar commented:
Then, there is nothing to test for me. Amazing @ybelenko, thank you!

ybelenko commented:
Check locally anyway. We spent huge amount of time with this, might be huge disappointment if it doesn't work for you.

nadar commented:
@ybelenko very hard to test i have to admit. Or is there any good command how i could make the openapi generator fetch the data from that branch?

openapi-generator-cli generate -i https://mysuperjson.json -g php --remote-generator=....... 

or can you give me a hint?

nadar commented:
@ybelenko i just copy the code quickly to my vendor files and adjust accordingly

```
$queryParams = array_merge($queryParams, ObjectSerializer::toQueryValue(
            $filter,
            'filter', // param base name
            'object', // openApiType
            'deepObject', // style
            true // explode
        ) ?? []);
```

And it works. :+1: Thanks

nadar commented:
great journey, thanks for the time @ybelenko

ksvirkou-hubspot commented:
Hey @ybelenko 
My app doesn’t work  because it past empty value to query string 
example bellow 
`https://api.*.com/crm/v3/objects/contacts?limit=10&after=&properties=website&properties`
Except string (without after)
`https://api.*.com/crm/v3/objects/contacts?limit=1&properties=website&properties`

wing328 commented:
Looks like this broke some of the PHP petstore integration tests:
```
PHPUnit 9.5.19 #StandWithUkraine

..............* Closing connection 0
.* Closing connection 0
................................................  63 / 124 ( 50%)
.........................................FF..........F.......   124 / 124 (100%)

Time: 00:01.443, Memory: 12.00 MB

There were 3 failures:

1) OpenAPI\Client\PetApiTest::testUpdatePetWithFormWithHttpInfo
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'updatePet'
+'update pet with form with http info'

/Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/tests/PetApiTest.php:165
phpvfscomposer:///Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/vendor/phpunit/phpunit/phpunit:97

2) OpenAPI\Client\PetApiTest::testUpdatePetWithForm
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'updatePet'
+'update pet with form'

/Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/tests/PetApiTest.php:178
phpvfscomposer:///Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/vendor/phpunit/phpunit/phpunit:97

3) OpenAPI\Client\RequestTest::testFormDataEncodingToJson
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'param=value&param2=value2'
+''

/Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/tests/RequestTest.php:34
phpvfscomposer:///Users/williamcheng/Code/openapi-generator3/samples/client/petstore/php/OpenAPIClient-php/vendor/phpunit/phpunit/phpunit:97
```

ybelenko commented on modules/openapi-generator/src/main/resources/php/api.mustache:
```diff
@@ -615,7 +597,7 @@ use {{invokerPackage}}\ObjectSerializer;
             } else {
                 // for HTTP post (form)
-                $httpBody = \GuzzleHttp\Psr7\Query::build($formParams);
+                $httpBody = ObjectSerializer::buildQuery($queryParams);
```
Should be `$formParams` instead of `$queryParams`. I broke it with a quick _find -> replace_ action in my IDE, sorry. I'm filling up new PR with a fix.
